### ISSUE TITLE
Clarity on AKS support for self hosted agents

### ISSUE THREAD

**Issue by ahmedsza** [microsoft/azure-pipelines-agent#3891]
This github issue - #3447  and associated article https://docs.microsoft.com/en-us/azure/devops/pipelines/agents/docker?view=azure-devops#use-azure-kubernetes-service-cluster mentions that the self hosted agent is only "supported" on AKS 1.18 or below. This is due to "docker in docker". It would be good to get additional clarification as to whether the self hosted agent is supported in AKS as long as docker and related functionality (like container jobs) are not used.. 

If I was to use a self hosted agent in AKS without the docker related functionality, would this be supported by Product Support?

I have also logged a docs issue - https://github.com/MicrosoftDocs/azure-devops-docs/issues/12455

**Comment by DmitriiBobreshev** [comment]
Hi @ahmedsza  thanks for reporting! We are working on more prioritized issues at the moment, but will get back to this one soon.

**Comment by leandromsft** [comment]
Hi @DmitriiBobreshev , I have the same question from my client. Please anyone clarify this issue? Thanks.

**Comment by Eldarrin** [comment]
Hi @Alexlofern , @leandromsft ,

I’ve added some new functionality to Keda using ado agents. Take a look here https://keda.sh/docs/2.8/scalers/azure-pipelines/ at the updated docs. This is a simple container calling the start.sh script from the MS self hosted agent page.

cheers, Eld

**Comment by Alexlofern** [comment]
Hi @Eldarrin, the major problem is the agent need docker in docker, after release 1.19, aks user containerd without support

https://learn.microsoft.com/en-us/azure/devops/pipelines/agents/docker?view=azure-devops#use-azure-kubernetes-service-cluster

**Comment by Alexlofern** [comment]
Hi @Eldarrin, how do you run azure devops agente withou DinD? https://learn.microsoft.com/en-us/azure/devops/pipelines/agents/docker?view=azure-devops#use-azure-kubernetes-service-cluster

**Comment by Eldarrin** [comment]
Hi @Alexlofern, you can run docker-in-docker without binding to the docker sock. You setup a PVC that will be the dind container disk, this caches and persists reboots, then a dind service to connect to it from another pod, then a dind container that is privileged but not socket bound. Then you access that via a docker network.

I have put my sample code here: https://github.com/Eldarrin/aks-dockerindocker (not tested it as I had to remove some company specific stuff) but should work

**Comment by lsprado** [comment]
Hi @Eldarrin 
I'm still confused 😢
self-hosted agent is supported on a kubernetes cluster above 1.18?

**Comment by Eldarrin** [comment]
HI @lsprado , yes. I am not sure if its official MSFT support but I am running the self-hosted agents on AKS 1.22.11

**Comment by leandromsft** [comment]
@DmitriiBobreshev do you have any update about this issue?

**Comment by leandromsft** [comment]
Hi @Eldarrin 
the question that in the documentation is saying that it's only supported on AKS versions 1.18 or below.
If I use AKS 1.22.11 like you, and I have a problem, will we ever get support from the product team?

@DmitriiBobreshev  any update about this question?

**Comment by ahmedsza** [comment]
@Eldarrin . You have this working with the docker in docker - i,e I can run docker build and use container jobs.. I looked at your repo - but was not sure on the steps to do this by myself. There seems to be 2 docker containers - one for the agent and then something else?

and importantly  - is this supported?

**Comment by Eldarrin** [comment]
This is what I have done to work around the removal of docker as the container system in AKS. My first solution was as detailed above. An independent docker in docker container using docker's own image and mounting a pvc to cache (pvc isn't required) and using a docker-network so that the agent uses docker's swarm network tech rather than trying to connect to the incluster docker.
I very much doubt this is supported as you may have to play with stuff like the docker MTU so the 2nd docker container can talk to some external repos (like jfrog/nexus).

I'm now converting all my docker pipelines to use kaniko as an in-cluster job in its own namespace and forgoing docker entirely.

**Comment by Eldarrin** [comment]
Just realised there is another way I had it working in development. Create a linux VM that has docker installed and set DOCKER_HOST in the agent and let the VM worry about the docker side of it

**Comment by luumarodrigues** [comment]
Hi @DmitriiBobreshev , do you already have a date to work in this question? I have the same question from my client and I would like to give him a more specific feedback. Please anyone to clarify this issue? 

Regards

**Comment by aabeing** [comment]
Contributing to this issue with some findings from my recent tests. I have experimented with the Kaniko approach and observed that it is less efficient compared to the traditional Docker build process.
Furthermore, integrating Kaniko within an AKS agent presents a complex challenge. This complexity not only makes the setup process difficult but also raises concerns about potential failures in the future.

In light of these findings, I propose an alternative solution. We could utilize standalone BuildKit daemon pods for our build process. It's worth noting that BuildKit is currently serving as the default builder for Docker. This approach could potentially streamline our build process and mitigate the issues we've encountered with Kaniko.
For reference explore the buildkit in kubernetes:
[https://github.com/moby/buildkit/tree/master/examples/kubernetes](https://github.com/moby/buildkit/tree/master/examples/kubernetes)
Further details on how we implemented this solution and resolved the limitations of AKS as self-hosted agents in Azure DevOps are described in the following blog.
[https://medium.com/@aabeing/aks-as-azure-devops-agents-buildkit-5af8e5cd43d1](https://medium.com/@aabeing/aks-as-azure-devops-agents-buildkit-5af8e5cd43d1)
Looking forward to your thoughts on this.
